extensions/sockets/src/main/java/org/mule/module/socket/internal/worker/TcpWorker.java

```diff
@@ -0,0 +1,246 @@
+                    {
+                        notify.wait();
+                    }
+                    catch (InterruptedException e)
+                    {
+                    }
+                }
+            }
+        }
+    }
+
+    private InputStream getNextMessage() throws IOException
+    {
+        InputStream readMsg = null;
+        try
+        {
+            readMsg = protocol.read(dataIn);
+            if (dataIn.isStreaming())
+            {
+                moreMessages.set(false);
+            }
+
+            return readMsg;
+        }
+        finally
+        {
+            if (readMsg == null)
+            {
+                dataIn.close();
+            }
+        }
+    }
+
+    protected void shutdownSocket() throws IOException
+    {
+        try
+        {
+            socket.shutdownOutput();
+        }
+        catch (UnsupportedOperationException e)
+        {
+            //Ignore, not supported by ssl sockets
+        }
+    }
+
+    private boolean hasMoreMessages()
+    {
+        return !socket.isClosed() && !dataInWorkFinished && moreMessages.get();
+    }
+
+    @Override
+    public void run()
+    {
+        while (hasMoreMessages())
+        {
+            InputStream content;
+            try
+            {
+                content = getNextMessage();
+            }
+            catch (IOException e)
+            {
+                if (LOGGER.isDebugEnabled())
+                {
+                    LOGGER.debug("An error occurred while reading from the TCP Worker connection", e);
+                }
+
+                moreMessages.set(false);
+                break;
+            }
+
+            if (content == null)
+            {
+                moreMessages.set(false);
+                break;
+            }
+
+            SocketAttributes attributes = new ImmutableSocketAttributes(socket);
+            messageHandler.handle(createMuleMessage(content, attributes, muleContext), new CompletionHandler<MuleEvent, Exception, MuleEvent>()
+            {
+                @Override
+                public void onCompletion(MuleEvent muleEvent, ExceptionCallback<MuleEvent, Exception> exceptionCallback)
+                {
+                    try
+                    {
+                        protocol.write(dataOut, muleEvent.getMessage().getPayload());
+                        dataOut.flush();
+                    }
+                    catch (IOException e)
+                    {
+                        exceptionCallback.onException(new IOException(
+                                format("An error occurred while sending TCP response to address '%s'",
+                                       socket.getRemoteSocketAddress().toString(), e))
+                        );
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e)
+                {
+                    // end worker's execution
+                    moreMessages.set(false);
```

Comment:
this doesn't look good. No logging of the exception? No log saying that you will actually stop the source? What if e is a ConnectionException? No notifying the exception callback?